traits/trait_types.py

```diff
@@ -4291,19 +4301,39 @@ def __init__(
     def validate(self, object, name, value):
         """ Check that the given value is valid date for this trait.
         """
-        if value is None and self.allow_none:
-            return value
-        if isinstance(value, datetime.date):
-            if self.allow_datetime or not isinstance(value, datetime.datetime):
+        if value is None:
+            if self.allow_none:
+                return value
+
+        elif isinstance(value, datetime.datetime):
+            if self.allow_datetime:
+                return value
+            elif self.allow_datetime is None:
+                warnings.warn(
+                    (
+                        "In the future, datetime.datetime instances will no "
+                        "longer be accepted by this trait type. To accept "
+                        "datetimes and silence this warning, use "
+                        "Date(allow_datetime=True) or Union(Datetime, Date)."
```

Comment:
Some subtlety here: `Union(Date, Datetime)` won't silence the warning, but `Union(Datetime, Date)` will.

Reply:
Hmm. Actually there's something funny going on with `Union` here: I don't get any warning when assigning a `datetime` instance to a `Union(Date, Float)` trait:

```
Python 3.9.2 (default, Feb 22 2021, 09:41:45) 
[Clang 12.0.0 (clang-1200.0.32.28)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> from traits.api import *
>>> class A(HasTraits):
...     foo = Union(Date, Float)
... 
>>> a = a90
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'a90' is not defined
>>> a = A()
>>> import datetime
>>> a.foo = datetime.datetime.today()
>>> ^D
```

Reply:
Okay, not so funny. This was because I was working at the prompt, which means that `DeprecationWarning`s other than those apparently coming from `__main__` are ignored. In the above case, the `DeprecationWarning` didn't appear to come from `__main__`, because I was using `stacklevel=2` when issuing the warning; that works for the plain old `Date` case, but not for `Union(Date, Float)` where there's an extra stack frame involved, so the right stack level is actually `3` rather than `2`. I don't see any reasonable way to know what the appropriate `stacklevel` to use is, so we're going to stick with `2`.

Reply:
Orthogonal: we should probably spell the trait type in the warning message as `Union(Datetime(), Date())` rather than `Union(Datetime, Date)`.